messages/src/message.rs

```diff
@@ -215,33 +169,16 @@ bitflags::bitflags! {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::is_valid_page;
-    use super::Error;
-    use super::MemoryRegion;
-
-    #[test]
-    fn test_is_valid_page() {
-        assert!(is_valid_page(0));
-        assert!(!is_valid_page(0x80));
-    }
-
-    #[test]
-    fn test_memory_region() {
-        let _ = MemoryRegion::new(0, 0, 10);
-        assert_eq!(
-            MemoryRegion::new(0x80, 0, 10).unwrap_err(),
-            Error::InvalidQsfpPage(0x80)
-        );
-        // Would read past map end.
-        assert!(matches!(
-            MemoryRegion::new(0x00, 1, 255).unwrap_err(),
-            Error::InvalidMemoryAccess { .. }
-        ));
-        assert!(matches!(
-            MemoryRegion::new(0x00, 255, 1).unwrap_err(),
-            Error::InvalidMemoryAccess { .. }
-        ));
-    }
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub enum PowerMode {
+    /// Entirely power off a transceiver module.
+    Off,
+    /// Request the module transition into low-power mode.
+    Low,
+    /// Request the module transition into high-power mode.
+    ///
+    /// Note that the actual power mode is defined in the transceiver's memory
+    /// map, in a location that depends on the exact management specification it
+    /// conforms to.
+    High,
```

Comment:
This API seems to leave a lot up to the SP.

For example, with an SFF-8636 device, we may need to set extra bits depending on class (class 5-7 need bit 2 in B93 set; class 8 need bit 3 in B93).

I'm confused, because it seems like we don't want the SP to have this much responsibility for parsing transceiver types.

Maybe the API is only intended to control the `LPMode` pin?  If so, we should give it a name that reflects the intention of controlling just the GPIO, instead of `PowerMode` more generally.

Reply:
> This API seems to leave a lot up to the SP.

That's fair, and definitely something we're trying to avoid! Yes, we don't want the SP mucking much with the memory map to determine the transceiver type.

I was expecting this to be effectively control of the LPMode pin, plus whatever the FPGA needs to do to cut the power entirely. If we wanted to set the actual power class, we'd do that in a preceding request. That may run afoul of the atomicity we're trying to achieve though. It might be that we need a compound message or something, which instructs the SP _how_ to set the power class, and also enable high-power mode? I.e., which bytes to write, and please de-assert LPMode for me.

Does that sound better? I think the message itself could just be a `MemoryRegion` inside that variant, but I may be missing something.

Reply:
My vague intuition would be to split "go to high-power mode" into two separate messages:
- Configure the relevant registers
- Deassert LPMode

(since the registers have no effect until the pin is low)

I guess I'm not sure what we're trying to protect against with the atomic request strategy.  As long as the host doesn't allow anyone else to inject messages midway through the "go to high power" process, I don't know who else would be modifying transceiver configuration.

Reply:
I've been looking at the specs in more detail, and you may be right that sending two messages is fine. I was under the impression that there may be a complex sequence of power modes that modules go through during initialization or something, and that the SDE would need to be sure that those have taken effect before deasserting `LPMode`. On closer reading, I see it's really about limiting the maximum power of a free-side module, in case some other, older component in the system would be damaged by the later power class 8, or the system isn't capable of supplying that power at all.

That is, we're generally just going to set byte 93 bits 2/3 for SFF-8636, and then deassert `LPMode` when we're ready. There's no real reason to expect that the SDE would separately twiddle those bits in a different way from the control plane. For CMIS modules, it doesn't look like there is a concept of limiting the power, just that there _is_ a high-power mode. (I could be missing something.)

You brought up earlier whether this API is intended to control just the `LPMode` pin. I still think we want to be able to completely power down a module, but I'm open to how we actually express that in the API. It seems like this would be sufficient, combined with a _separate_ write to 93[2-3] where needed (SFF-8636 modules). Does that sound reasonable? Or would you prefer separate messages for controlling the `LPMode` pin and completely powering down a module?

As a related note, I think we'll need to update the status bitflags I've already got. There's no way to express a module being present, but completely powered down, I think. Does that sound right?

Reply:
Ah, no my point about the status flags is not correct. The `Status::ENABLED` bit is outside the spec, and tells us whether there is power at all delivered to the module, independent of whether low- or high-power mode is enabled. Nevermind!

Reply:
Thanks for the details!  After thinking a little more, I think this `enum` would be fine with some additional commenting, basically saying

> - In `Off` mode, we cut power to the transceiver using the hot-swap / e-fuse
> - In `Low` mode, we enable power to the transceiver but keep `LPMode` high.  Please do not set `Power_override`, as this defeats the purpose of controlling this in hardware!
> - In `High` mode, we enable power to the transceiver and pull `LPMode` low; further configuration of registers may be required to actually enter high-power mode (depending on device class)

Reply:
Thanks, that sounds good! I believe I've addressed this in https://github.com/oxidecomputer/transceiver-control/pull/5/commits/4575e0e1220151a1208842c81bba667cf6dd64fe, but let me know if you have anything outstanding.

Reply:
FWIW, when I originally wrote this, my assumption was that the host would configure any of the actual registers itself before asking the SP to the release the low power pin. So my expectation was that the onus was not on the SP.

Reply:
Yep, understood! I think @mkeeter and I agree on that point as well. The bulk of this discussion was about whether writing the relevant registers and releasing into high-power mode needed to be a single message from host to SP. I think we arrived at "no", at least for the time being.

Part of the reason is that the host library I'm writing will include a function to enable high-power mode. That will use two exchanges under the hood, to set the correct registers if appropriate, and then deassert `LPMode`, but we _think_ that should be OK at this point. Let me know if you think that's _not_ the case @rmustacc. In the meantime, I'll make that assumption clear in the documentation.

Reply:
I believe that should be fine. What I was trying to avoid was what seemed like at one point an additional command that was specific to this process of setting the i2c registers.